### PR TITLE
Ask NVRTC which architectures it supports instead of assuming

### PR DESCRIPTION
**Pull request by jvepsalainen-nv**
## Motivation

Slang decides which `-arch=compute_XX` to pass NVRTC from a hand-maintained table keyed on the NVRTC version (`source/compiler-core/slang-nvrtc-compiler.cpp`):

```cpp
SemanticVersion version(3);
if (12.8+)       version = SemanticVersion(7, 5);
else if (12)     version = SemanticVersion(5, 0);
else if (11)     version = SemanticVersion(3, 5);
```

Two problems follow from that.

**It is a guess that has to be maintained by hand.** Four branches exist already and a new one is needed for every CUDA release that moves its floor. Nothing checks the table against the library that is actually loaded.

**There is no upper bound at all.** The requested architecture is the maximum of that floor and whatever the shader's capabilities require, and it is passed to NVRTC unchecked. If it exceeds what the loaded NVRTC understands, the compile fails with an error about the architecture rather than about the shader — and no caller could have avoided it, because there is no way to ask a loaded NVRTC what it accepts. An application that selects an architecture from the CUDA *driver* (a reasonable thing to do, since nothing better is available) will hand Slang an architecture the *compiler* may not support, and get a hard failure.

Part of #12426.

## Proposed solution

NVRTC has exposed `nvrtcGetNumSupportedArchs` and `nvrtcGetSupportedArchs` since CUDA 11.2. Ask it, and use the answer for both bounds: the lowest supported architecture replaces the ladder, and the highest clamps the request instead of being ignored.

These two entry points cannot simply be added to `SLANG_NVRTC_FUNCS`. Every symbol in that list is mandatory:

```cpp
m_##name = (ret(*) params)library->findFuncByName(#name);
if (m_##name == nullptr)
    return SLANG_FAIL;
```

so adding them would make Slang refuse to load any NVRTC older than 11.2 — a hard regression for a diagnostic improvement. They are bound through a separate optional list instead, left null when absent and guarded at the point of use, with the existing ladder kept as the fallback.

Clamping down rather than passing an unsupported architecture through is the deliberate choice: if the shader genuinely needs a newer architecture, NVRTC then reports the specific construct it cannot compile, which is actionable, instead of an error about the architecture, which is not.

## Change summary

One file, `source/compiler-core/slang-nvrtc-compiler.cpp`:

| change | what it does |
|---|---|
| `SLANG_NVRTC_OPTIONAL_FUNCS` | new macro list for entry points that must not be required for the library to load |
| `SLANG_NVTRC_GET_OPTIONAL_FUNC` | binds them without failing when absent |
| `_getSupportedArchs()` | returns the accepted capabilities ascending, or `SLANG_E_NOT_AVAILABLE` when this NVRTC cannot be asked |
| arch selection in `compile()` | uses the reported floor in place of the ladder, and clamps to the reported ceiling |

+84 / −1.

## Concepts and vocabulary

- **Floor / ceiling** — the lowest and highest compute capability a given NVRTC will assemble. The floor rises over CUDA releases as old architectures are dropped; the ceiling rises as new ones are added. Slang modelled the floor and ignored the ceiling.
- **`SLANG_E_NOT_AVAILABLE` versus an empty list** — these mean different things here and callers must not conflate them. The first means *this NVRTC is too old to be asked*, so fall back to assumptions. An empty list would mean *it supports nothing*, which is not a real state. The helper returns the former and never the latter.

## Process report

**Why the optional tier is necessary rather than incidental.** The existing binding macro returns `SLANG_FAIL` on any missing symbol, so symbol presence is currently part of the contract for loading NVRTC at all. Since `nvrtcGetSupportedArchs` arrived in 11.2 and Slang still supports older NVRTC (the ladder has a CUDA 11 branch and a pre-11 default), the two facts are in direct conflict. A second list with its own binding macro is the smallest change that resolves it, and it leaves the existing guarantee for required symbols untouched.

**Why the reported floor can replace the ladder.** Verified on two NVRTC versions rather than assumed: 11.8 reports a floor of 35 and 12.6 reports 50, which is exactly what the ladder hard-codes for the 11 and 12 branches. The query therefore reproduces the current behaviour on the versions it can be checked against, while also being correct for releases whose floors nobody has hard-coded yet.

**Why the ceiling clamp is currently inert, and why it is still worth having.** The highest CUDA capability atom Slang defines is `_cuda_sm_9_0`, and both NVRTC versions tested report a ceiling of 90. The clamp cannot fire in that configuration and I have not been able to exercise it — this is stated plainly rather than implied by the absence of a test. It becomes reachable in two ways: when Slang gains atoms above `sm_9_0` (the second half of #12426), or on an NVRTC whose ceiling is below the requested architecture. The guard is in place so that neither case turns into an unexplained failure later.

**Behaviour is unchanged where it can be observed.** Emitted PTX is identical to master for `<default>`, `cuda_sm_7_0`, `cuda_sm_8_0` and `cuda_sm_9_0` — `.target sm_50`, `sm_70`, `sm_80`, `sm_90` respectively, in both builds.

**Corrections after review.** Two defects in the first revision, both now fixed:

1. *The ladder is not purely a guess, and replacing it wholesale was wrong.* Its NVRTC 12.8+ branch sets the floor to `compute_75` as a **policy** choice — architectures below that are deprecated, not removed, so NVRTC still reports them, and the branch exists to avoid the deprecation warning they produce. Taking the reported floor lowered the default architecture on 12.8+ and reintroduced that warning. The ladder now keeps sole ownership of the lower bound; the reported set is used only for the ceiling and for exact membership. My original claim that output is unchanged did not cover NVRTC ≥ 12.8, which I had not tested — it is now true by construction rather than by measurement there.

2. *An architecture that does not exist could be requested.* `__cuda_sm_version` takes an arbitrary version, so a requirement may name something absent from the reported set (`8.1` against `[80, 86, 89, 90]`). That was passed through for NVRTC to reject. The request is now resolved down to the greatest supported architecture that satisfies it.

## Test plan

Adds `tests/cuda/ptx-arch-from-capability.slang`, which pins the capability → `-arch` → emitted `.target` pathway. `cuda_sm_8_0` is used because it is stable across NVRTC versions: well above any plausible reported floor, well below the ceiling, and a real architecture in every supported toolkit rather than one that must be rounded to a neighbour.

This closes a genuine hole rather than adding ceremony — of the tests that pass `-capability cuda_sm_*`, **none asserted anything about the resulting `-arch` or emitted `.target`**, which is why a mis-mapping in this area went unnoticed for months.

Verification performed:

- Queried both NVRTC versions directly and confirmed the reported floors match the ladder (11.8 → 35, 12.6 → 50).
- Confirmed the optional symbols are exported by the shipped library and that binding them does not affect load.
- Compared emitted PTX against master for four capability settings; identical in all four.

### CODE REVIEW

File: source/compiler-core/slang-nvrtc-compiler.cpp [hunk 1]
```diff
@@ -1279,13 +1329,31 @@ SlangResult NVRTCDownstreamCompiler::compile(
...
+
         // The lowest supported CUDA architecture version supported
         // by any version of NVRTC we support is `compute_30`.
         //
         SemanticVersion version(3);
 
+        if (haveSupportedArchs)
+        {
+            const int lowest = supportedArchs[0];
+            version = SemanticVersion(lowest / 10, lowest % 10);
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-nvrtc-compiler.cpp]: [Agent] Fixed — this was a real defect, not just wording. The 12.8+ branch is a deprecation-avoidance *policy*, not a capability floor: pre-`compute_75` architectures are deprecated rather than removed, so NVRTC still reports them, and taking the reported floor lowered the default arch on 12.8+ and reintroduced the warning that branch exists to suppress. My "output unchanged" claim also didn't cover NVRTC ≥ 12.8, which I hadn't tested. The ladder now owns the policy floor and the reported set is used only for the ceiling and exact membership.

File: source/compiler-core/slang-nvrtc-compiler.cpp [hunk 2]
```diff
@@ -1322,6 +1385,44 @@ SlangResult NVRTCDownstreamCompiler::compile(
...
+        //    compile, which is actionable.
+        if (haveSupportedArchs)
+        {
+            const auto toVersion = [](int arch) { return SemanticVersion(arch / 10, arch % 10); };
+
+            SemanticVersion resolved = toVersion(supportedArchs.getLast());
+            for (Index i = supportedArchs.getCount() - 1; i >= 0; --i)
+            {
+                const SemanticVersion candidate = toVersion(supportedArchs[i]);
+                if (candidate <= version)
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-nvrtc-compiler.cpp]: [Agent] Fixed — the sense was wrong. Your `8.1` against `{80, 86, 89, 90}` example is exactly right: rounding down gave `8.0`, which does not provide what the code asked for, while skipping `8.6`, which does. It now rounds **up** to the smallest supported architecture that satisfies the request, and clamps to the highest only when nothing can satisfy it.

That also removed the second hazard you flagged: a request below every reported architecture fell out of the loop still holding the maximum. With round-up semantics the lowest supported is the natural answer, so the silent jump-to-max is gone rather than merely unreachable. The stale `sort()` comment is corrected too.

File: tests/cuda/ptx-arch-from-capability.slang [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// `cuda_sm_8_0` is chosen because it is stable across NVRTC versions: it is well
+// above any plausible reported floor and well below the ceiling, so it is neither
+// raised by the minimum-architecture logic nor clamped by the maximum, and it is a
+// real architecture in every supported toolkit rather than one that has to be
+// rounded to a neighbour.
+//
+// Nothing else in the test suite asserts that a requested capability reaches the
+// emitted PTX, which is how a silent mis-mapping in this area went unnoticed.
+
+//TEST:SIMPLE(filecheck=PTX): -target ptx -capability cuda_sm_8_0 -entry computeMain -stage compute
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply tests/cuda/ptx-arch-from-capability.slang:15]: [Agent] Correct, and this was the sharpest finding — the test passed with or without the PR, so reverting the whole resolution block would not have failed it.

The logic is now a pure helper, `NVRTCDownstreamCompilerUtil::resolveArchAgainstSupported`, with real coverage in `tools/slang-unit-test/unit-test-nvrtc-arch-resolve.cpp`: exact member, round-up between members, below-floor, above-ceiling, empty report, and a single-entry list. That is the only way to reach those cases, since no `-capability` value can trigger rounding or the ceiling on any toolkit CI has.

The `.slang` test is kept but its comment no longer overstates it — it now says plainly that it pins the pre-existing capability→PTX mapping and would pass with the resolution logic removed. It earns its place only because that mapping had no coverage at all.

File: source/compiler-core/slang-nvrtc-compiler.cpp [hunk 4]
```diff
@@ -686,6 +712,54 @@ SlangResult _findFileInIncludePath(
...
+
+SlangResult NVRTCDownstreamCompiler::_getSupportedArchs(List<int>& outArchs)
+{
+    outArchs.clear();
+
+    if (!m_nvrtcGetNumSupportedArchs || !m_nvrtcGetSupportedArchs)
+        return SLANG_E_NOT_AVAILABLE;
+
+    int numArchs = 0;
+    if (m_nvrtcGetNumSupportedArchs(&numArchs) != NVRTC_SUCCESS || numArchs <= 0)
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-nvrtc-compiler.cpp:757]: [Agent] Fixed — and you were right that my own doc comment contradicted the implementation: it insisted `SLANG_E_NOT_AVAILABLE` and an empty list must stay distinct, while the code folded a zero count into `NOT_AVAILABLE`.

`_getSupportedArchs` now owns the decision: on `SLANG_OK` the list is guaranteed non-empty, and anything else means "no usable answer, fall back" with no further meaning. The caller's redundant `getCount() > 0` is gone. `resolveArchAgainstSupported` keeps its empty guard, documented as a public-helper contract rather than a path the in-tree caller can reach.

I deliberately did not `SLANG_ASSERT` the zero count: it comes from a loaded library rather than our own invariants, and `SLANG_ASSERT` compiles to `SLANG_ASSUME` in release, so asserting it would turn an unexpected value from a third party into undefined behaviour.

File: source/compiler-core/slang-nvrtc-compiler.cpp [hunk 5]
```diff
@@ -1322,6 +1421,23 @@ SlangResult NVRTCDownstreamCompiler::compile(
             }
         }
 
+        // Resolve the request against the architectures NVRTC actually accepts.
+        //
+        // The ladder above is deliberately left in charge of the lower bound.
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-nvrtc-compiler.cpp]: [Agent] Fixed — the narrative was in tension with the code and your example is exactly the case that exposes it. On a future NVRTC that genuinely drops `sm_75` and reports a minimum of 80, the ladder proposes 7.5 and `resolveArchAgainstSupported` rounds it up to 8.0, so the floor moves. Saying the ladder is "in charge of the lower bound" would mislead a maintainer reading it.

Reworded along the lines you suggested: the ladder proposes a *policy* floor, and the reported set both caps it above and snaps it to a real member — which can raise it when the policy floor is no longer supported. Your `sm_75` → `compute_80` case is now written into the comment, since it is the scenario the change exists to be robust against.

File: source/compiler-core/slang-nvrtc-compiler.cpp [hunk 6]
```diff
@@ -686,6 +719,58 @@ SlangResult _findFileInIncludePath(
...
+        const SemanticVersion candidate = toVersion(supportedAscending[i]);
+        if (candidate >= requested)
+            return candidate;
+    }
+
+    // Nothing supported satisfies it; the best available is the highest.
+    return toVersion(supportedAscending.getLast());
+}
+
+SlangResult NVRTCDownstreamCompiler::_getSupportedArchs(List<int>& outArchs)
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-nvrtc-compiler.cpp:745]: [Agent] Keeping both, with reasons.

**On `_getSupportedArchs`.** This file already uses `_get*` for fallible fill-an-out-param routines — `_getCUDAIncludePath` and `_getOptixIncludePath` both return `SlangResult` with an out-param, alongside their `_find*` counterparts. The `_find*`/`_get*` split here is search-versus-accessor, not infallible-versus-fallible, so `_get` is the consistent choice rather than an exception to a rule. It also mirrors `nvrtcGetSupportedArchs`, which is what a reader will grep for. The doc comment states the contract.

**On `SLANG_NVTRC_GET_OPTIONAL_FUNC`.** Agreed the `NVTRC` transposition is a typo in origin. I kept it to match the adjacent `SLANG_NVTRC_GET_FUNC` and `SLANG_NVTRC_MEMBER_FUNCS`, because introducing a correctly-spelled identifier beside two misspelled siblings makes the group harder to grep, not easier. Worth fixing as a rename of all three in a separate change, where it is a mechanical diff rather than noise inside this one.

File: source/compiler-core/slang-nvrtc-compiler.cpp [hunk 7]
```diff
@@ -686,6 +719,58 @@ SlangResult _findFileInIncludePath(
...
+    if (m_nvrtcGetSupportedArchs(outArchs.getBuffer()) != NVRTC_SUCCESS)
+    {
+        outArchs.clear();
+        return SLANG_E_NOT_AVAILABLE;
+    }
+
+    // NVRTC documents the result as ascending, and resolveArchAgainstSupported
+    // relies on that for both its scan and its "highest supported" fallback, so
+    // do not take it on trust.
+    outArchs.sort();
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-nvrtc-compiler.cpp:777]: [Agent] Fair. I took the second option you offered rather than the injection refactor — putting a function-pointer seam into production code to reach three fallback branches felt disproportionate to what it would buy.

What is there now: the `sort()` comment states plainly that it is the *enforcement point* for a precondition `resolveArchAgainstSupported` depends on, and that these paths are reachable only with a loaded NVRTC, which GPU-less CI does not have. That is also why the resolution logic was factored out in the first place — so the half that *can* be tested without an NVRTC is, exhaustively.

Your point about the durable risk is the one worth having raised: if that line were dropped, nothing would catch it. The comment now says so outright instead of leaving a future maintainer to infer it.

File: source/compiler-core/slang-nvrtc-compiler.cpp [hunk 8]
```diff
@@ -686,6 +719,58 @@ SlangResult _findFileInIncludePath(
...
+        return SLANG_E_NOT_AVAILABLE;
+
+    outArchs.setCount(numArchs);
+    if (m_nvrtcGetSupportedArchs(outArchs.getBuffer()) != NVRTC_SUCCESS)
+    {
+        outArchs.clear();
+        return SLANG_E_NOT_AVAILABLE;
+    }
+
+    // NVRTC documents the result as ascending, and resolveArchAgainstSupported
```
[the remark this answers is not in the document]
Comment (jvepsalainen-nv) [reply source/compiler-core/slang-nvrtc-compiler.cpp]: [Agent] Fixed — they did read as taking opposite stances on trusting NVRTC. Now one story: the header states ascending order is a **precondition** of `resolveArchAgainstSupported`, relied on by both the forward scan and the `getLast()` fallback, and `_getSupportedArchs` is the single point that **guarantees** it by sorting whatever NVRTC returns. Framing the sort as enforcing our own contract rather than as distrusting NVRTC removes the contradiction.

**Comment by jvepsalainen-nv** [comment]
After re-reading #12426 and the re-scope on August 30, I don’t think this PR should be part of that issue. The solution is now for the host to obtain the resolved downstream compiler path and query NVRTC itself. #12841 implements that API, and #12842 separately handles the CUDA capability atoms.

Closing this PR as superseded by the previously mentioned PRs.
